### PR TITLE
define the GicTrait by crate_interface

### PR DESCRIPTION
**Pull request by luodeb**

### CODE REVIEW

File: src/gic_v2.rs [hunk 1]
```diff
@@ -444,3 +444,67 @@ impl GicCpuInterface {
         self.regs().PMR.set(0xff);
     }
 }
+
+// define the GicTrait by crate_interface
```
Comment (hky1999) [inline src/gic_v2.rs]: `///`

File: src/gic_v2.rs [hunk 2]
```diff
@@ -444,3 +444,67 @@ impl GicCpuInterface {
...
+#[crate_interface::def_interface]
+pub trait GicTrait {
```
Comment (aarkegz) [inline src/gic_v2.rs:452]: doc comments for each methods?

File: src/lib.rs [hunk 3]
```diff
@@ -7,7 +7,10 @@ use core::ops::Range;
 
 mod gic_v2;
 
-pub use gic_v2::{GicCpuInterface, GicDistributor};
+/// Since the implementation of `crate_interface` requires fully 
+/// exporting `GicTrait`, otherwise `crate_interface` will not 
+/// be able to recognize it.
+pub use gic_v2::{GicCpuInterface, GicDistributor, GicTrait, GicInetrface};
```
Comment (aarkegz) [inline src/lib.rs]: misspelled `GicInetrface`?

**Review by aarkegz — CHANGES_REQUESTED** [review]

File: src/gic_v2.rs [hunk 4]
```diff
@@ -444,3 +444,67 @@ impl GicCpuInterface {
...
+    fn set_icfgr(int_id: usize, cfg: u8);
+
+    fn get_target_cpu(int_id: usize) -> usize;
+    fn set_target_cpu(int_id: usize, target: u8);
+
+    fn get_priority(int_id: usize) -> usize;
+    fn set_priority(int_id: usize, priority: u8);
+}
+
+/// Implementation of [`GicTrait`] by crate_interface::call_interface,
```
Comment (hky1999) [inline src/gic_v2.rs:484]: This is extremely strange.

~~You should `impl GicTrait for GicInterface`~~

~~and use methods provied by `GicInterface` for the implementation of `GicTrait`~~

Why did you call `call_interface!(xxx)` here...

It should by called by vgic, maybe?

**Review by hky1999 — CHANGES_REQUESTED** [review]

**Comment by hky1999** [comment]
Basic idea:
* `GicTrait` defined and provided by `arm_gicv2`
* `GicTrait` implemented by arceos driver
* `GicTrait` imported and called by `arm_vgic`

File: src/gic_v2.rs (hunk 4 above)
Comment (hky1999) [reply src/gic_v2.rs:484]: > This is extremely strange.
> 
> ~You should `impl GicTrait for GicInterface`~
> 
> ~and use methods provied by `GicInterface` for the implementation of `GicTrait`~
> 
> Why did you call `call_interface!(xxx)` here...
> 
> It should by called by vgic, maybe?

Sorry for mis-understanding, `GicTrait` should implemented somewhere with the **GIC instance**, which is hold by arceos

Comment (luodeb) [reply src/gic_v2.rs:484]: We added another layer of abstraction on top of the original Trait, which has two benefits:  
1. It avoids exposing all the contents of `gic_v2`, such as `gic_v2::*`.  
2. We can simply use `GicInterface::set_enable()` without needing to import unnecessary `crate_interface`, achieving minimal dependency.

File: src/lib.rs (hunk 3 above)
Comment (luodeb) [reply src/lib.rs]: ok

File: src/gic_v2.rs (hunk 4 above)
Comment (hky1999) [reply src/gic_v2.rs:484]: > We added another layer of abstraction on top of the original Trait, which has two benefits:
> 
> 1. It avoids exposing all the contents of `gic_v2`, such as `gic_v2::*`.
> 2. We can simply use `GicInterface::set_enable()` without needing to import unnecessary `crate_interface`, achieving minimal dependency.

I have no problem with that,

`arm_vgic` can import `GicInterface` provided by `arm_gicv2` directlty.

**Comment by pengzechen** [comment]
https://github.com/arceos-hypervisor/arceos/pull/26
axhal的实现在这里 @hky1999

File: src/gic_v2.rs (hunk 2 above)
Comment (hky1999) [reply src/gic_v2.rs:452]: @pengzechen mind this comment please

**Review by hky1999 — CHANGES_REQUESTED** [review]

**Review by hky1999 — APPROVED** [review]

**Pull request merged**